src/diff/children.js

```diff
@@ -46,14 +45,15 @@ export function diffChildren(parentDom, newParentVNode, oldParentVNode, context,
 		// Check if we find a corresponding element in oldChildren and store the
 		// index where the element was found.
 		p = oldChildren[i];
-		if (p != null && (childVNode.key==null && p.key==null ? (childVNode.type === p.type) : (childVNode.key === p.key))) {
-			index = i;
-		}
-		else {
-			for (j=0; j<oldChildrenLength; j++) {
-				p = oldChildren[j];
-				if (p!=null) {
-					if (childVNode.key==null && p.key==null ? (childVNode.type === p.type) : (childVNode.key === p.key)) {
+
+		if (childVNode!=null) {
+			if (p != null && (childVNode.key==null && p.key==null ? (childVNode.type === p.type) : (childVNode.key === p.key))) {
+				index = i;
+			}
+			else if (childVNode.key!=null) {
+				for (j=0; j<oldChildrenLength; j++) {
+					p = oldChildren[j];
+					if (p!=null && childVNode.key === p.key) {
```

Comment:
The `<Input>` component is not a good example because the state is bound to the dom node and there is no real component state. A true class component like in the linked codesandbox better shows the issue.

Maybe we need to check if the component is a true component class and only then skip matching by type? Just thinking out loud.

Reply:
The more I think about it, the more I like to always re-use functional components. That should make the `<Input />` example work without keys. They can be safely swapped out without leading to an incorrect render result because they don't have any instance variables or component state. Both approaches definitely have their strengths and weaknesses.

Currently the mentioned example will break in Preact: `condition ? [<C1 />, <C2 />] : [<C2 />, <C2 />]`. Let's imagine the following actions happening:

1. render `<C1 />, <C2 />`
2. user updates state in `<C2 />`
3. another update higher up is triggered and the condition is now `false` -> `<C2 />, <C2 />`
3.1. During reconciliation our current solution (without this PR) will see that the components at the first position are not equal (`C1 !== C2`) and will iterate over the children array to find a matching `vnode`.
3.2. The second `C2` which holds the state the user previously updated does match that criteria and our reconciler will use it for the first element.
3.3. Since we've moved the second element to the first position we need to recreate the second component from scratch.

What effectively happens is that the second component has lost any state or custom instance variables.

Reply:
Thinking some more about it the same is true for functional components that have hooks. Hooks typically add state to a component and reusing them may also lead to incorrect results.

Reply:
After going back and forth on this a few times, I'm convinced that my original stance has flaws. Even better, all this can be side-tracked by just checking for holes explicitly. I apologize for not seeing it sooner.